### PR TITLE
Fix SVG image rendering issue in MHTML

### PR DESCRIPTION
**Pull request by devKshitijJain**
Fixes - https://github.com/shihabmridha/educative.io-downloader/issues/28

| Before | After |
|:-:|:-:|
| Screenshot 2020-05-06 at 4 17 47 AM | Screenshot 2020-05-06 at 4 17 34 AM |

### CODE REVIEW

File: src/download.ts [hunk 1]
```diff
@@ -188,19 +188,42 @@ async function downloadPage(title: string, link: string): Promise<void> {
...
+  // Convert SVG image into Data URL and  also get rid of unescape characters
+  const parentElements = document.getElementsByClassName('canvas-svg-viewmode');
+  for (let i = 0; i < parentElements.length; i++) {
+    const parentElement = parentElements[i];
+    const svgElement = parentElement.getElementsByTagName('svg')[0];
+
+    const imageTags = svgElement.getElementsByTagName('image');
+    for (let j = 0; j < imageTags.length; j++) {
+      const imageTag = imageTags[j];
+      const blob = await fetch(imageTag.getAttribute('xlink:href')).then(r => r.blob());
```
Comment (shihabmridha) [inline src/download.ts:214]: Can you try like this:
```suggestion
      const response = await fetch(imageTag.getAttribute('xlink:href'));
      const blob = response.blob();
```

File: src/download.ts [hunk 2]
```diff
@@ -188,19 +188,42 @@ async function downloadPage(title: string, link: string): Promise<void> {
...
+      let dataUrl = await new Promise(resolve => {
+        let reader = new FileReader();
```
Comment (shihabmridha) [inline src/download.ts]: Lets use `const`.

**Comment by the-redback** [comment]
Good to see fast response. (y)
I tested this PR. Works fast enough with `headless: false`. But, with, `headless: true`, it takes something around 100s to load a single page that contains SVG. It would be great help if you could do some magic regarding itthis. :D
This is an example page: https://www.educative.io/courses/grokking-computer-networking/3jQgm8xAnxA

**Review by shihabmridha — CHANGES_REQUESTED** [review]
Right now, you are requesting one url at a time. We should send multiple concurrent requests to boost the speed a bit up.
I am gonna merge this anyway but this should be done.

**Comment by devKshitijJain** [comment]
Well that's something new. Pull Bot closed it. Gonna reopen it.

File: src/download.ts (hunk 1 above)
Comment (devKshitijJain) [reply src/download.ts:214]: what would this do?  @shihabmridha

File: src/download.ts (hunk 2 above)
Comment (devKshitijJain) [reply src/download.ts]: @shihabmridha Done

**Comment by devKshitijJain** [comment]
@the-redback well image has to be loaded and then it need to be read using fileReader to get data uri. I am not sure how I can improve it.

**Comment by devKshitijJain** [comment]
> Right now, you are requesting one url at a time. We should send multiple concurrent requests to boost the speed a bit up.
> I am gonna merge this anyway but this should be done.

Well I am not sure about it. The thing is even if we put it in array of promise. If 2nd image loads faster than first, what will happen in that case?

File: src/download.ts (hunk 1 above)
Comment (devKshitijJain) [reply src/download.ts:214]: blob will have to be await as well. Isn't is same as doing something in then or using await?

Comment (shihabmridha) [reply src/download.ts:214]: This results the same but more readable format.

**Comment by the-redback** [comment]
> @the-redback well image has to be loaded and then it need to be read using fileReader to get data uri. I am not sure how I can improve it.

Yes. And, it takes only 1/2 seconds to do all of this in `headless: false`. But, `headless: true` mode is so slow that each iterate takes 2/3 seconds, in total 100s. I wonder if it is possible to make it as faster as `headless: false`.

File: src/download.ts (hunk 1 above)
Comment (devKshitijJain) [reply src/download.ts:214]: okay

**Comment by devKshitijJain** [comment]
My assumption is that when headless is false in that case, the image is already visible and in memory, thus it takes very little time.

**Review by shihabmridha — APPROVED** [review]

**Pull request merged**